### ISSUE TITLE
False positive E1133 (not-an-iterable) when returning (type hinted) class variable

### ISSUE THREAD

**Issue by Flamefire** [pylint-dev/pylint#10298]
### Bug description

```python
When returning a class variable from an instance the type is not correctly detected and E1133 is raised although a list will be returned.

To show I use 2 classes where one returns an instance variable and the other a class variable. The annotation is only shown for the class variable:


# pylint: disable=missing-module-docstring, missing-function-docstring, missing-class-docstring, invalid-name, attribute-defined-outside-init

class MyClass:
    @classmethod
    def setUpClass(cls):
        MyClass.x = None

    @classmethod
    def fill(cls):
        MyClass.x = [1]

    def value(self) -> list[int]:
        if MyClass.x is None:
            MyClass.fill()
        return MyClass.x


class MyClass2:
    def fill(self):
        self.x = [1]

    def value(self) -> list[int]:
        if self.x is None:
            self.fill()
        return self.x


for i in MyClass().value():  # Warning here
    print(i)

for i in MyClass2().value():  # No warning here
    print(i)


This might be related to https://github.com/pylint-dev/pylint/issues/6044 which also uses a class variable and/or https://github.com/pylint-dev/pylint/issues/2766 which tries to solve a similar issue with type hints.
```

### Configuration

```ini

```

### Command used

```shell
pylint --enable=E1133 /tmp/a.py
```

### Pylint output

```python
************* Module a
/tmp/a.py:28:9: E1133: Non-iterable value MyClass().value() is used in an iterating context (not-an-iterable)
```

### Expected behavior

No warning show for either case and/or at least being able to silence it using standard Python type hints as shown.

### Pylint version

```shell
pylint 3.3.6
astroid 3.3.8
Python 3.12.1 (main, Dec 20 2023, 10:28:56) [GCC 11.4.0]
```

### Additional dependencies

```python

```
